pinot-broker/src/test/java/org/apache/pinot/broker/queryquota/HelixExternalViewBasedQueryQuotaManagerTest.java

```diff
@@ -531,25 +674,41 @@ private static ExternalView generateBrokerResource(String tableName) {
   private void runQueries()
       throws InterruptedException {
     runQueries(TABLE_MAX_QPS, false);
-    //increase the qps and some of the queries should be throttled.
-    runQueries(TABLE_MAX_QPS * 2, true);
+    // increase the qps and some of the queries should be throttled.
+    // keep in mind that permits are 'regenerated' on every call based on how much time elapsed since last one
+    // that means for 25 QPS we get new permit every 40 ms or 0.5 every 20 ms
+    // if we start with 25 permits at time t1 then if we want to exceed the qps in the next second  we've to do more
+    // double requests, because 25 will regenerate
+    runQueries(TABLE_MAX_QPS * 2 + 1, true);
+  }
+
+  private void runQueries(double qps, boolean shouldFail)
+      throws InterruptedException {
+    runQueries(qps, shouldFail, APP_NAME);
   }
 
   // try to keep the qps below 50 to ensure that the time lost between 2 query runs on top of the sleepMillis
   // is not comparable to sleepMillis, else the actual qps would end being lot lower than required qps
-  private void runQueries(double qps, boolean shouldFail)
+  private void runQueries(double qps, boolean shouldFail, String appName)
       throws InterruptedException {
     int failCount = 0;
     long sleepMillis = (long) (1000 / qps);
     for (int i = 0; i < qps; i++) {
+      if (!_queryQuotaManager.acquireApplication(appName)) {
+        failCount++;
+      }
       if (!_queryQuotaManager.acquireDatabase(CommonConstants.DEFAULT_DATABASE)) {
         failCount++;
       }
       if (!_queryQuotaManager.acquire(RAW_TABLE_NAME)) {
         failCount++;
       }
-      Thread.sleep(sleepMillis);
```

Comment:
I replaced sleeps with timesource.advanceByMillis(x) to make test more stable (because sleeping could be affected by GC or random interrupts) and much faster. I'd be cool if similar approach could be used in integration test.

Reply:
Any improvement here would be great, but I think they should be discussed in another PR, not here. It is not clear to me how this proposed system works and I don't think we should block the application level quota in order to discuss about the timesource method

Reply:
Sure, that note is informational. I'm not suggesting to improve ITests in this PR.

Reply:
Just to be clear. I'm super into adding something like this in another PR. Using Thread.sleep is a very bad practice. But it is not clear to me how this api you introduced works and I think we can do that in another PR.

Reply:
The second change is a small/simple improvement of test code and doesn't change how rate limiters behave. 
Instead of sleeping in tests (which is unreliable), we advance the time reported by time source injected into rate limiter. 
In the end rate limiter reads that variable instead of System.nanoTime(). 
I don't think it's big enough to warrant putting it into separate PR.

Reply:
Removed.